## M345SC Lab 2

The course bitbucket repo has been updated a few times in the last week. Ensure that you can sync your online fork of the repo. If/when using git on MLC machines, you should first launch *tortoise git* from the software hub, and then you will be provided with git options when right-clicking on a file or folder

#### Task 1: Timing code

There are a few different approaches for timing code in Python.
You can use the *time* function in the time module as below:

In [1]:
import numpy as np
import time
X = list(np.random.randint(1,1000,800)) #uncomment and use if/as needed
t1 = time.time()
y = np.sin(np.outer(X,X))
t2 = time.time()
print("dt = ",t2-t1)

dt =  0.02901172637939453


And you can also use the timeit function:

In [2]:
timeit np.sin(np.outer(X,X))

28.1 ms ± 515 µs per loop (mean ± std. dev. of 7 runs, 10 loops each)


You will now investigate how the runing time for sorting algorithms depends on the list length. Functions for merge and insertion sorts are below.

In [3]:
def isort(X):
    """Sort X using insertion sort algorithm and return sorted array
    """

    S = X.copy()

    for i,x in enumerate(X[1:],1):
        #place x appropriately in partially sorted array, S
        for j in range(i-1,-1,-1):
            if S[j+1]<S[j]:
                S[j],S[j+1] = S[j+1],S[j]
            else:
                break
    return S

def merge(L,R):
    """Merge 2 sorted lists provided as input
    into a single sorted list
    """
    M = [] #Merged list, initially empty
    indL,indR = 0,0 #start indices
    nL,nR = len(L),len(R)

    #Add one element to M per iteration until an entire sublist
    #has been added
    for i in range(nL+nR):
        if L[indL]<R[indR]:
            M.append(L[indL])
            indL = indL + 1
            if indL>=nL:
                M.extend(R[indR:])
                break
        else:
            M.append(R[indR])
            indR = indR + 1
            if indR>=nR:
                M.extend(L[indL:])
                break
    return M

def mergesort(X):
    """Given a unsorted list, sort list using
    merge sort algorithm and return sorted list
    """

    n = len(X)

    if n==1:
        return X
    else:
        L = mergesort(X[:n//2])
        R = mergesort(X[n//2:])
        return merge(L,R)

1) How much does the run-time for insertion sort increase when the length of the initial, unsorted list is doubled? What do you expect?

In [103]:
N = 2**11
X = list(np.random.randint(1,2*N,N))
t1 = time.time()
isort(X)
t2 = time.time()
N = N * 2
X = list(np.random.randint(1,2*N,N))
t3 = time.time()
isort(X)
t4 = time.time()
print("dt1 = ",t2-t1)
print("dt2 = ",t4-t3)

dt1 =  0.6953189373016357
dt2 =  2.619417905807495


2) Now carry out the analagous analysis for mergesort.

In [116]:
N = 2**11
X = list(np.random.randint(1,2*N,N))
t1 = time.time()
mergesort(X)
t2 = time.time()
N = N * 2
X = list(np.random.randint(1,2*N,N))
t3 = time.time()
mergesort(X)
t4 = time.time()
print("dt1 = ",t2-t1)
print("dt2 = ",t4-t3)

dt1 =  0.007872343063354492
dt2 =  0.0327601432800293


3) Consider one *level* of mergesort where *m* sorted lists of length *p* are merged into *m/2* lists of length *2p* (assume that *m* is even), and we have $N = mp$. In lecture, the running time for one level was stated to be $O(N)$. This indicates the number of required operations is: $cN + f(N)$ with $lim_{n\rightarrow \infty} f/N = 0$ and $c$ a constant independent of N. Construct an (useful) upper bound for $c$. Does this bound depend on *m* or *p*?

#### Answer

m sorted lists of length p so total length will be mp

merge into m/2 lists of length 2p so total length will be mp

Total length mp = N

TO DO

####  Task 2: Setting up a word count
In this task, you will place the words from a text file into a Python dictionary so that word counts for specified words can be quickly extracted. The line below will read in the file (if the notebook and the text file, *words.txt*, are in the same directory).

In [28]:
infile = open('words.txt','r')
A = infile.read()
#Remove punctuation
A = A.replace(',','')
A = A.replace('?','')
A = A.replace('.','')
A = A.replace('!','')
A = A.lower() #Convert to all lowercase
words = A.split() #List (mostly) containing the words in the file
print(words[:4])
infile.close()

['sing', 'michael', 'sing', 'on']


1) Fill the dictionary below so that each word in *words* is stored as a key and the corresponding value is the number of times the word appears in the list. If you are unfamiliar with dictionaries in Python, have a look at the online tutorial: https://docs.python.org/3/tutorial/datastructures.html#dictionaries

In [102]:
D = {} #Initializes empty dictionary
#Add code below
wordsorted = mergesort(words)
ln = len(wordsorted)
value = 1
for i in range(1,ln-1):
    if wordsorted[i] == wordsorted[i-1]:
        value += 1
    else:
        D[wordsorted[i-1]] = value
        value = 1

2) Using your dictionary, look up how many times does the word "fail" appears in the file? (you can check your result by applying the *count* method on *words*)

In [121]:
timeit D['fail']

92.8 ns ± 0.715 ns per loop (mean ± std. dev. of 7 runs, 10000000 loops each)


3) In general (i.e. not specifically for this file) what are the running times for a) creating the dictionary and b) looking up a word count in the dictionary?

#### Answer
The running times for creating a dictionary are: 72.9 ns ± 0.796 ns per loop

The running times to look up a word count in the dictionary: 92.8ns ± 0.715 ns per loop